src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/SymbolKey/SymbolKey.NamedTypeSymbolKey.cs

```diff
@@ -71,13 +76,25 @@ public static SymbolKeyResolution Resolve(SymbolKeyReader reader, out string? fa
             private static void Resolve(
                 PooledArrayBuilder<INamedTypeSymbol> result,
                 INamespaceOrTypeSymbol container,
-                string metadataName,
+                string name,
                 int arity,
+                string? filePath,
                 bool isUnboundGenericType,
                 ITypeSymbol[] typeArguments)
             {
-                foreach (var type in container.GetTypeMembers(GetName(metadataName), arity))
+                foreach (var type in container.GetTypeMembers(name, arity))
                 {
+                    // if this is a 'file' type, then only resolve to a file-type from this same file
+                    if (filePath != null)
```

Comment:
If you start with a non-file symbol, can this still resolve to a file symbol that's unambiguious? Is that intentional?